### ISSUE TITLE
Idea for a different method for flagging warnings, linked to improved localization and false positive handling

### ISSUE THREAD

**Issue by DavidWikstrom** [jwillinghalpern/fm_rainbow_log#18]
So this is something that has been at the back of my head while looking at the code and thinking about the various issues. I'll try to just give a quick abstract of my ideas:

- It might be beneficial to not (only) explicitly check for warnings/errors, but also and first of all check for strings that we know are NOT errors or warnings but indicates operations that are successes. That should avoid a false positive like with the import from a file name "ya existe.fmp12". Anything that is not a know success or an error will then be either a known warning or an unhandled warning. This does require setting up an additional set of sample actions to get the relevant strings for each language.

- It could be advantageous to detect and store the language of the current line(s) while parsing. That would allow for future options like giving user feedback in the language of the log line, and, perhaps more importantly, some of the parsing code could be segmented into different sections per language (possibly even in separate files). This should increase code readability, may make it easier to have specific parsing logic for certain languages, and would facilitate letting contributors work on the specific code for the languages they master.

- Error codes: this is a bit more ambitious and possibly trickier, but if one were to create individual error codes for each known error/warning (and a generic one for those not explicitly handled), it should then be possible to customize colors/notifications/sounds per error code. This could help address one challenge I see with notifications right now, is that they currently also get triggered by actions that don't actually require any action on part of the developer (like pasting an extra "If" in a script).

**Comment by jwillinghalpern** [comment]
tldr: I think I'm most inclined to work on suggestion 2 to separate language definitions into different files. That would help improve the legibility of the codebase, since it's a wall of text right now.

> It might be beneficial to not (only) explicitly check for warnings/errors

This sounds reasonable, but I'm not 100% sold on "Anything that is not a know success or an error will then be either a known warning or an unhandled warning." It seems that could create false warning positives unless fmrl encodes every possible success string.

I added a false positives test which we should expand first with more reasonable potential false positives in `false-positives.txt`. Then only take remedial steps if false positives seem likely. Please contribute some if any come to mind. With cargo installed, you can modify false-positives.txt and run either `cargo test avoid_false_positives` or to keep the test running `cargo watch --exec "test avoid_false_positives"`.
<img width="922" alt="Screenshot 2023-05-05 at 10 23 59 AM" src="https://user-images.githubusercontent.com/16845102/236525529-18afe3fe-40b2-4a47-ae5f-2df8ad7121e7.png">


> detect and store the language...some of the parsing code could be segmented into different sections per language

I really like the idea of separate files, mainly for readability and to make it more contribution-friendly. I'm not sure the language has to be detected for each line, though. I think we could still brute force and check every language for every line. I suppose it doesn't hurt to theoretically speed up parsing, but it hasn't bit us yet.

> Error codes:

I like this too. I have a vague idea about how to separate the language files by defining a trait that describes _everything_ a proper LineParser must do, then implement a LineParser for each language. That trait contract could also require the assignment of an an errorcode (a predetermined enum) for each parsed line. Then those enums could control notification settings and maybe even be exposed to the config.json file for user customization.

**Comment by DavidWikstrom** [comment]
Some great updates you've released lately!

Success strings do indeed have the limitation that they only are a good idea if you know all success strings, and at this point, I'm not certain whether that is "easy" to do or not. My plan is to try to execute all the actions I can come up with that would getting to Import.log, very much like what we did for errors, and then I'll look at the data and see if this seems like a fruitful path to further explore.
I am still very much inclined to think that handling of success strings would be very helpful in eliminating false positive; that could also be implemented without going to the extreme of considering anything that is not a known success an error/warning.

**Comment by jwillinghalpern** [comment]
I wonder if this issue is adequately addressed by the new `error_rules` feature?

**Comment by DavidWikstrom** [comment]
I was also just going to say that you can close this issue now.

**Comment by jwillinghalpern** [comment]
https://github.com/jwillinghalpern/fm_rainbow_log/issues/26
